edb/schema/schema.py

```diff
@@ -818,21 +906,49 @@ def add(
 
         return self._replace(**updates)  # type: ignore
 
+    def add(
+        self,
+        id: uuid.UUID,
+        sclass: Type[so.Object],
+        data: Mapping[str, Any],
+    ) -> FlatSchema:
+        reducible_fields = sclass.get_reducible_fields()
```

Comment:
What's going on here?

Reply:
Basically we don't want non-primitive Python object instances in the schema anymore.  And so, if you have a field that has a type `schema.objects.Object` or `objects.ObjectCollection`, etc, we call `schema_reduce` on it to get a primitive tuple.  `schema_restore` reverses the process and is called when you call the corresponding getter.  Specifically, objects are reduced to a `(classname, id)` tuple, and object collections are reduced to a `(classname, typeargs, tuple<id>, tuple<attr, val>)`.